Rework profiler argument handling.

kumpera opened the pull request.

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,406 @@
+#endif
+
+
+#include "mono-profiler-log.h"
```
Already included above.

kumpera replied:
Fixed

alexrp commented on mono/profiler/mono-profiler-log.h:
```diff
@@ -153,4 +156,108 @@ enum {
+#define MAX_FRAMES 32
+
+typedef enum {
+	DomainEvents = 1 << 0,
```
Is there a reason for this enum naming style? It deviates from the style we use throughout most of the runtime AFAIK. I was actually a bit confused what `DomainEvents` was when I was reading the arg-parsing.c code above. My first thought was "what is a type name doing here?".

kumpera replied:
No reason at all.
I'm replacing it with PROFLOG_ defines.

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,406 @@
+	{ "counter", CounterEvents },
+	{ "gchandle", GCHandleEvents },
+
+	{ "typesystem", TypeLoadingAlias },
```
Naming inconsistency here.

kumpera replied:
Hmmmmmm
Perhaps so, but that gets us to long boring names.
I rather rename it to something like types, TBH.

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,406 @@
+static void
+load_args_from_env_or_default (ProfilerConfig *config)
+{
+	//XXX change this to header constants
```
Would that serve an actual purpose? They're just default values after all.

kumpera replied:
Consistency with the rest of the runtime.

All constants are better thrown together in a conf header. Similar to sgen-conf.h

It allow us to have long form description/discussion of their values.

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,406 @@
+	printf ("\tsample[=TYPE]        enable/disable statistical sampling of threads (by default cycles/100)\n");
+	printf ("\t                     TYPE: cycles,instr,cacherefs,cachemiss,branches,branchmiss\n");
+	printf ("\t                     TYPE can be followed by /FREQUENCY\n");
+	printf ("\t[heapshot[=MODE]     record heap shot info (by default at each major collection)\n");
```
Too many `[`s.

kumpera replied:
Fixed.

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,406 @@
+	printf ("\t[no]calls            enable/disable recording enter/leave method events\n");
+	printf ("\t[no]legacy           enable/disable pre mono 5.4 default profiler events\n");
+	printf ("\tsample[=TYPE]        enable/disable statistical sampling of threads (by default cycles/100)\n");
+	printf ("\t                     TYPE: cycles,instr,cacherefs,cachemiss,branches,branchmiss\n");
```
We should only document `cycles`. The others were part of the Linux perf support which we've removed.

alexrp replied:
Actually, maybe we should take this opportunity to just get rid of the old syntax (in `set_sample_freq`) and make it `sample=$freq` only.

kumpera replied:
Let's remove TYPE.
I'm inclined to allow cycles/100 to reduce the transition pain for the profiler team.
Actually. @rodrmoya, does the profiler passes `cycles/100` or just `100`?

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,406 @@
+	printf ("\t[no]coverage         enable collection of code coverage data\n");
+	printf ("\tcovfilter=ASSEMBLY   add an assembly to the code coverage filters\n");
+	printf ("\t                     add a + to include the assembly or a - to exclude it\n");
+	printf ("\t                     filter=-mscorlib\n");
```
`covfilter`

kumpera replied:
fixed

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,406 @@
+	printf ("\tcovfilter-file=FILE  use FILE to generate the list of assemblies to be filtered\n");
+	printf ("\tmaxframes=NUM        collect up to NUM stack frames\n");
+	printf ("\tcalldepth=NUM        ignore method events for call chain depth bigger than NUM\n");
+	printf ("\toutput=FILENAME      write the data to file FILENAME (-FILENAME to overwrite)\n");
```
We always overwrite now, so we should probably take this opportunity to drop support for `-FILENAME`, or at least deprecate it.

kumpera replied:
Removed the docs and printing a warning when used.
I agree we should change the code to always unlink before opening.

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,406 @@
+	{ "gcalloc", AllocationEvents },
+	{ "gc", GCEvents },
+	{ "thread", ThreadEvents },
+	{ "calls", EnterLeaveEvents }, //calls is the old name, we can just keep it
```
I definitely prefer `calls`.

kumpera replied:
Comment removed then :)

alexrp commented on mono/profiler/mono-profiler-log.c:
```diff
@@ -4791,217 +4578,52 @@ mono_profiler_startup_log (const char *desc)
 	mono_profiler_startup (desc);
 }
 
+static ProfilerConfig config;
```
It would be nice to move this near the top of the file and use it instead of all the global variables we have right now. Not a blocker though.

alexrp commented on mono/profiler/mono-profiler-log.c:
```diff
@@ -5011,34 +4633,106 @@ mono_profiler_startup (const char *desc)
+		mono_profiler_install_gc_moves (gc_moves);
+	}
+
+	if (config.effective_mask & (GCRootEvents | GCHandleEvents)) {
```
Minor thing: It would be nice to add a `TODO` here since we should really split these events into two separate flags in the profiler API.

kumpera replied:
added

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,406 @@
+	config->sample_freq = 100;
+	config->max_call_depth = 100;
+	config->num_frames = MAX_FRAMES;
+	config->debug_coverage |= g_hasenv ("MONO_PROFILER_DEBUG_COVERAGE");
```
I just realized this is the only case of the profiler using environment variables, and I can't really think of a good reason why. Would be nice to turn this into an option. Can be addressed later though.

kumpera replied:
I added the option, I'll remove the env var. It's for internal usage, so no big deal.

If really needed, we can add something like MONO_ENV_OPTIONS.

alexrp commented on mono/profiler/mono-profiler-log.c:
```diff
@@ -4791,217 +4578,52 @@ mono_profiler_startup_log (const char *desc)
+	if (desc [3] == ':')
+		proflog_parse_args (&config, desc + 4);
+	else
+		proflog_parse_args (&config, "");
```
`proflog_parse_args (&config, desc [3] == ':' ? desc + 4 : "")` ?

kumpera replied:
done

vkargov commented on mono/profiler/mono-profiler-log.c:
```diff
@@ -3838,6 +3732,32 @@ init_suppressed_assemblies (void)
+	FILE *filter_file;
+	char *line, *content;
+
+	if (filters == NULL)
```
`filters` are already initialized in line 4614. Furthermore, it's a function-local variable, so the assigned array will be lost once we leave the function

kumpera replied:
good catch!

vkargov commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,406 @@
+	const char *p;
+	gboolean in_quotes = FALSE;
+	char quote_char = '\0';
+	char *buffer = malloc (strlen (desc));
```
The buffer is never freed?

kumpera replied:
Fixed

alexrp commented:
By the way, I'm not sure how I feel about `arg-parsing.c` as file name when the main file is called `mono-profiler-log.c` to mirror others (e.g. `mono-profiler-aot.c`). It's not immediately obvious from the file name which profiler the file belongs to. Maybe we should move the log profiler as a whole into a `log` subdirectory?

kumpera commented on mono/profiler/mono-profiler-log.c:
```diff
@@ -5011,34 +4633,106 @@ mono_profiler_startup (const char *desc)
 	init_thread (prof, TRUE);
 
+	//This two events are required for the profiler to work
+	int events = MONO_PROFILE_THREADS | MONO_PROFILE_GC;
```
We should, I haven't come around doing it. Will do it next.

kumpera commented:
Pushed changes addressing most comments in this PR. Thanks for the review!

Still missing/TODO:

- Sort out the fate of arg-parsing.c. The name sucks for the state reasons but I don't think we need to mono the log profiler to a new directory for now.

- Filtering mandatory events from MLPD files.
- Split the gc roots/ gc handles events in the profiler backend
- Better fix for -FILENAME

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -297,19 +302,18 @@ usage (void)
-	printf ("\tsample[=TYPE]        enable/disable statistical sampling of threads (by default cycles/100)\n");
-	printf ("\t                     TYPE: cycles,instr,cacherefs,cachemiss,branches,branchmiss\n");
+	printf ("\tsample[=frequency]   enable/disable statistical sampling of threads (frequency in Hz, 100 by default)\n");
 	printf ("\t                     TYPE can be followed by /FREQUENCY\n");
```
This line should be removed.

alexrp commented on mono/profiler/mono-profiler-log.c:
```diff
@@ -4485,6 +4485,7 @@ create_profiler (const char *args, const char *filename, GPtrArray *filters)
 	if (filename && *filename == '-') {
 		force_delete = 1;
 		filename++;
+		printf ("WARNING: the output:-FILENAME option is deprecated, the profiler now always overrides the output file\n");
```
`g_warning` so it doesn't get lost on Android/iOS?

kumpera replied:
Done

alexrp commented on mono/profiler/arg-parsing.c:
```diff
@@ -0,0 +1,408 @@
+	printf ("\t[no]calls            enable/disable recording enter/leave method events\n");
+	printf ("\t[no]legacy           enable/disable pre mono 5.4 default profiler events\n");
+	printf ("\tsample[=frequency]   enable/disable statistical sampling of threads (frequency in Hz, 100 by default)\n");
+	printf ("\t                     TYPE can be followed by /FREQUENCY\n");
```
This line should be removed.

alexrp commented:
One option for the file names would be to shorten them to `log.c`, `aot.c`, etc and then have `log-args.c`. Those names would be a bit nicer to type.
